### PR TITLE
feat(api): move-to-slot JSON protocol command

### PR DESCRIPTION
**Pull request by IanLondon**
## overview

This an an experimental feature intended to allow users to control pipette position without reference to aspirating and dispensing, and without reference to labware/modules. Slot seems like a reasonable reference to use, especially since Slot 1's origin coincides with the origin of the deck. This should be generally useful for applications from colony picking to potential future deck calibration move-to's.

For details of the `move-to-slot` command params, see the new JSON schema v2 file (called out in a comment below).

The `move-to-slot` command can be marked as **non-experimental** once we move past APIv1, which limits control of the `move_to` command. But as-is, the behavior is different in APIv1 vs APIv2.

This addition necessitates a new JSON schema, version 2. PD will stay at JSON Protocol Schema v1, because it doesn't read commands and doesn't need to write `move-to-slot` commands anytime soon. I expect we'll jump PD from JSON protocol v1 right to v3 with embedded labware defs and other major changes. I also expect we'll carry over this `move-to-slot` command to JSON protocol schema v3 by camel-casing the keys and not changing anything else with the command.

## changelog

* add args to `InstrumentContext.move_to` and `plan_moves` to allow more control of motion
* new JSON protocol schema v2 - only addition is new `move-to-slot` command
* support `move-to-slot` command in APIv1 and APIv2 JSON executors

## review requests

Here is a test file [moveToTest.json.zip](https://github.com/Opentrons/opentrons/files/2994075/moveToTest.json.zip) - to run in APIv2, change the labware on L61 from `"96-flat"` to `"generic_96_wellplate_380_ul"`. 

- [ ] Review JSON schema with [JSON Schema Validator](http://jsonschemavalidator.net/) (test the `move-to-slot` command, it should fail validation on nonsense inputs and pass on good inputs)
- [ ] New `"description"` text is JSON is sufficiently readable as documentation
- [ ] Python API code review
- [ ] Test on robot with API v1, with listed limitations
- [ ] Test on robot with API v2, `force-direct` and `minimum-z-height` should have full functionality

### CODE REVIEW

File: shared-data/protocol-json-schema/protocolSchemaV2.json [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+                    },
+                    "params": {
+                      "allOf": [
+                        {"$ref": "#/definitions/pipette-access-params"}
+                      ]
+                    }
+                  }
+                },
+
+                {
```
Comment (IanLondon) [inline shared-data/protocol-json-schema/protocolSchemaV2.json]: this is the only new section

File: api/src/opentrons/protocol_api/contexts.py [hunk 2]
```diff
@@ -1231,18 +1231,21 @@ def _execute_transfer(self, plan: transfers.TransferPlan):
...
-                         on the labware from which and to which the pipette is
-                         moving; if it is 0, the pipette will move directly;
-                         and if it is non-zero, the pipette will rise to the
-                         z_safety point before moving in x and y.
+        :param force_direct: If set to true, move directly to destination
+                        without arc motion.
+        :param minimum_z_height: An optional height to retract the pipette to
+                         before moving. If not specified, it will be generated
+                         based on the labware from which and to which the
+                         pipette is moving; if it is 0, the pipette will move
```
Comment (sfoster1) [inline api/src/opentrons/protocol_api/contexts.py]: This comment isn't accurate anymore I think

File: api/src/opentrons/protocol_api/execute.py [hunk 3]
```diff
@@ -312,6 +312,24 @@ def dispatch_json(context: ProtocolContext,  # noqa(C901)
             well = _get_well(labware, params)
             offset = default_values.get('touch-tip-mm-from-top', -1)
             pipette.touch_tip(location, v_offset=offset)  # type: ignore
+
+        elif command_type == 'move-to-slot':
+            slot = params.get('slot')
+            if slot not in [str(s+1) for s in range(12)]:
```
Comment (sfoster1) [inline api/src/opentrons/protocol_api/execute.py]: Can't we do `slot in context.deck`

File: shared-data/protocol-json-schema/protocolSchemaV2.json [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+        "p10_multi",
+        "p50_single",
+        "p50_multi",
+        "p300_single",
+        "p300_multi",
+        "p1000_single",
+        "p1000_multi"
+      ]
+    },
+    "pipette-model": {
```
Comment (sfoster1) [inline shared-data/protocol-json-schema/protocolSchemaV2.json]: If this is v2 can we remove this

Comment (IanLondon) [reply shared-data/protocol-json-schema/protocolSchemaV2.json]: I was hoping for this bump we could limit changes to only adding the new command, and do everything else we're planning together in #3110

File: api/src/opentrons/protocol_api/execute.py (hunk 3 above)
Comment (IanLondon) [reply api/src/opentrons/protocol_api/execute.py]: I think that's too permissive b/c it will work fine with integers even though the schema says it must be a string: `"enum": ["1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12"]`. 

But I'm pretty sure `1 in context.deck` and `"1" in context.deck` will both be True... didn't test though

Probably the best thing to do is to throw before execution even begins if the protocol does not conform to a schema supported by the executor, but until we have something like that I don't want users to run out-of-spec JSON and have things work fine - then they'll not bother making it a string and we'll have a tough time making a migration script to move their protocols to future JSON schema versions :/

Comment (sfoster1) [reply api/src/opentrons/protocol_api/execute.py]: Yeah I agree with that but still, "what the deck will accept" is exactly the source of truth we should be checking here, not duplicating logic.

File: shared-data/protocol-json-schema/protocolSchemaV2.json (hunk 4 above)
Comment (sfoster1) [reply shared-data/protocol-json-schema/protocolSchemaV2.json]: 👍

File: api/src/opentrons/protocol_api/execute.py (hunk 3 above)
Comment (b-cooper) [reply api/src/opentrons/protocol_api/execute.py]: I agree with Seth about duplicating logic and to extend that thought, this would be a great place to start using the ot2Standard.json deck definition in shared data. And a helper method to grab the slot ids from the definition would return an array of strings by default.

Comment (IanLondon) [reply api/src/opentrons/protocol_api/execute.py]: The source of truth for the JSON executor about what inputs are valid should always be the JSON Protocol schema def, not the Deck class of a certain version of the Python API, or anything else outside the JSON Protocol schema def matching the protocol version of the uploaded protocol.

Permissivity in the executors is dangerous. Our JSON Protocol schema is not intended to be read only by the JSON executor(s) in our Python server/API. For example, we might want to display commands info in Run App or Protocol Library from the JSON file directly. If we encourage protocols to deviate from the schema definition by having a permissive executor, our JSON protocols "in the wild" will become much harder to work in all these cases. As a superuser, if my protocol runs fine when I use integers for slots, I'm not going to really care if I'm going against the schema -- I likely won't even know that my protocol is invalid because it runs fine on the robot. Then if Run App wants to add a new feature that reads the `slots` param and expects a string, we have to either make users convert their invalid protocols to string slots, or handle that out-of-spec case to the application.

I think the long-term right thing to do is adding [JSON schema validation](https://github.com/Julian/jsonschema) checks to the executors, which would throw an error if the uploaded protocol does not validate under the JSON Protocol schema that matches that uploaded protocol's schema version (`"protocol-schema" / "protocolSchema"` key). We should probably do that pretty soon. Then, we can take this type of "is the input valid?" assertions out of the executors.

Comment (IanLondon) [reply api/src/opentrons/protocol_api/execute.py]: #3250

File: api/tests/opentrons/protocol_api/test_geometry.py [hunk 5]
```diff
@@ -5,8 +5,8 @@
 from opentrons.protocol_api import labware
 from opentrons.hardware_control.types import CriticalPoint
 
-labware_name = 'generic_96_wellplate_380_ul'
-trough_name = 'usa_scientific_12_trough_22_ml'
+labware_name = 'generic_96_wellPlate_380_uL'
```
Comment (mcous) [inline api/tests/opentrons/protocol_api/test_geometry.py]: Can you revert these changes? All lower-case is now the correct thing to do
